File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        int n = arr.size(); 
        if(n <= 1) return 1; 
        
        map<int,int> dp; 
        int max_ = 1; 
        for(int i = 0; i < n; i++){
            int target = arr[i] - difference; 
            if(dp.count(target)){
                dp[arr[i]] = dp[target] + 1; 
                max_ = max(max_,dp[arr[i]]); 
            } else{
                dp[arr[i]] = 1; 
            }
        }
        return max_; 
    }
};

File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.java
class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        int n = arr.length; 
        if(n <= 1) return 1; 
        Map<Integer,Integer> dp = new HashMap<>(); 
        int max_ = 1; 
        
        for(int i = 0; i < n; i++){
            int target = arr[i] - difference; 
            if(dp.containsKey(target)){
                dp.put(arr[i], dp.getOrDefault(target,0) + 1); 
                max_ = Math.max(max_, dp.get(arr[i])); 
            } else{
                dp.put(arr[i],1); 
            }
        }

        return max_; 
    }
}